File: python/src/ensoul/resources/simulations.py
"""Simulations resource for the Ensoul SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ensoul.generated.simulations import SimulationDetailResponse

if TYPE_CHECKING:
    from ensoul.http import AsyncHTTPClient, SyncHTTPClient
    from ensoul.pagination import AsyncPage, SyncPage
    from ensoul.streaming import AsyncSSEStream, SyncSSEStream

__all__ = [
    "Simulations",
    "AsyncSimulations",
]


class Simulations:
    """Synchronous simulations resource."""

    def __init__(self, client: SyncHTTPClient) -> None:
        self._client = client

    def create(
        self,
        *,
        name: str,
        domain_id: str,
        description: str | None = None,
        config: dict[str, Any] | None = None,
        participant_persona_ids: list[str] | None = None,
    ) -> SimulationDetailResponse:
        """POST /v1/simulations"""
        body: dict[str, Any] = {"name": name, "domain_id": domain_id}
        if description is not None:
            body["description"] = description
        if config is not None:
            body["config"] = config
        if participant_persona_ids is not None:
            body["participant_persona_ids"] = participant_persona_ids
        response = self._client.post("/v1/simulations", json=body)
        return SimulationDetailResponse.model_validate(response.json())

    def get(self, simulation_id: str) -> SimulationDetailResponse:
        """GET /v1/simulations/{simulation_id}"""
        response = self._client.get(f"/v1/simulations/{simulation_id}")
        return SimulationDetailResponse.model_validate(response.json())

    def list(
        self, *, page: int = 1, per_page: int = 20, **kwargs: Any
    ) -> SyncPage[dict]:
        """GET /v1/simulations"""
        from ensoul.pagination import SyncPage

        params: dict[str, Any] = {"page": page, "per_page": per_page}
        params.update({k: v for k, v in kwargs.items() if v is not None})
        response = self._client.get("/v1/simulations", params=params)
        data = response.json()
        return SyncPage(
            items=data["items"],
            total=data["total"],
            page=data["page"],
            per_page=data["per_page"],
            pages=data["pages"],
            client=self._client,
            method="GET",
            path="/v1/simulations",
            params=params,
            model=dict,
        )

    def start(
        self, simulation_id: str, *, ticks: int | None = None
    ) -> dict[str, Any]:
        """POST /v1/simulations/{simulation_id}/start"""
        body: dict[str, Any] = {}
        if ticks is not None:
            body["ticks"] = ticks
        response = self._client.post(
            f"/v1/simulations/{simulation_id}/start", json=body
        )
        return response.json()

    def pause(self, simulation_id: str) -> dict[str, Any]:
        """POST /v1/simulations/{simulation_id}/pause"""
        response = self._client.post(
            f"/v1/simulations/{simulation_id}/pause", json={}
        )
        return response.json()

    def stop(self, simulation_id: str) -> dict[str, Any]:
        """POST /v1/simulations/{simulation_id}/stop"""
        response = self._client.post(
            f"/v1/simulations/{simulation_id}/stop", json={}
        )
        return response.json()

    def stream(self, simulation_id: str) -> SyncSSEStream:
        """GET /v1/simulations/{simulation_id}/stream"""
        return self._client.stream_sse(
            "GET", f"/v1/simulations/{simulation_id}/stream"
        )

    def get_events(
        self,
        simulation_id: str,
        *,
        page: int = 1,
        per_page: int = 20,
        **kwargs: Any,
    ) -> SyncPage[dict]:
        """GET /v1/simulations/{simulation_id}/events"""
        from ensoul.pagination import SyncPage

        params: dict[str, Any] = {"page": page, "per_page": per_page}
        params.update({k: v for k, v in kwargs.items() if v is not None})
        response = self._client.get(
            f"/v1/simulations/{simulation_id}/events", params=params
        )
        data = response.json()
        return SyncPage(
            items=data["items"],
            total=data["total"],
            page=data["page"],
            per_page=data["per_page"],
            pages=data["pages"],
            client=self._client,
            method="GET",
            path=f"/v1/simulations/{simulation_id}/events",
            params=params,
            model=dict,
        )

    def get_history(self, simulation_id: str) -> dict[str, Any]:
        """GET /v1/simulations/{simulation_id}/history"""
        response = self._client.get(f"/v1/simulations/{simulation_id}/history")
        return response.json()


class AsyncSimulations:
    """Async version of the simulations resource."""

    def __init__(self, client: AsyncHTTPClient) -> None:
        self._client = client

    async def create(
        self,
        *,
        name: str,
        domain_id: str,
        description: str | None = None,
        config: dict[str, Any] | None = None,
        participant_persona_ids: list[str] | None = None,
    ) -> SimulationDetailResponse:
        """POST /v1/simulations"""
        body: dict[str, Any] = {"name": name, "domain_id": domain_id}
        if description is not None:
            body["description"] = description
        if config is not None:
            body["config"] = config
        if participant_persona_ids is not None:
            body["participant_persona_ids"] = participant_persona_ids
        response = await self._client.post("/v1/simulations", json=body)
        return SimulationDetailResponse.model_validate(response.json())

    async def get(self, simulation_id: str) -> SimulationDetailResponse:
        """GET /v1/simulations/{simulation_id}"""
        response = await self._client.get(f"/v1/simulations/{simulation_id}")
        return SimulationDetailResponse.model_validate(response.json())

    async def list(
        self, *, page: int = 1, per_page: int = 20, **kwargs: Any
    ) -> AsyncPage[dict]:
        """GET /v1/simulations"""
        from ensoul.pagination import AsyncPage

        params: dict[str, Any] = {"page": page, "per_page": per_page}
        params.update({k: v for k, v in kwargs.items() if v is not None})
        response = await self._client.get("/v1/simulations", params=params)
        data = response.json()
        return AsyncPage(
            items=data["items"],
            total=data["total"],
            page=data["page"],
            per_page=data["per_page"],
            pages=data["pages"],
            client=self._client,
            method="GET",
            path="/v1/simulations",
            params=params,
            model=dict,
        )

    async def start(
        self, simulation_id: str, *, ticks: int | None = None
    ) -> dict[str, Any]:
        """POST /v1/simulations/{simulation_id}/start"""
        body: dict[str, Any] = {}
        if ticks is not None:
            body["ticks"] = ticks
        response = await self._client.post(
            f"/v1/simulations/{simulation_id}/start", json=body
        )
        return response.json()

    async def pause(self, simulation_id: str) -> dict[str, Any]:
        """POST /v1/simulations/{simulation_id}/pause"""
        response = await self._client.post(
            f"/v1/simulations/{simulation_id}/pause", json={}
        )
        return response.json()

    async def stop(self, simulation_id: str) -> dict[str, Any]:
        """POST /v1/simulations/{simulation_id}/stop"""
        response = await self._client.post(
            f"/v1/simulations/{simulation_id}/stop", json={}
        )
        return response.json()

    async def stream(self, simulation_id: str) -> AsyncSSEStream:
        """GET /v1/simulations/{simulation_id}/stream"""
        return await self._client.stream_sse(
            "GET", f"/v1/simulations/{simulation_id}/stream"
        )

    async def get_events(
        self,
        simulation_id: str,
        *,
        page: int = 1,
        per_page: int = 20,
        **kwargs: Any,
    ) -> AsyncPage[dict]:
        """GET /v1/simulations/{simulation_id}/events"""
        from ensoul.pagination import AsyncPage

        params: dict[str, Any] = {"page": page, "per_page": per_page}
        params.update({k: v for k, v in kwargs.items() if v is not None})
        response = await self._client.get(
            f"/v1/simulations/{simulation_id}/events", params=params
        )
        data = response.json()
        return AsyncPage(
            items=data["items"],
            total=data["total"],
            page=data["page"],
            per_page=data["per_page"],
            pages=data["pages"],
            client=self._client,
            method="GET",
            path=f"/v1/simulations/{simulation_id}/events",
            params=params,
            model=dict,
        )

    async def get_history(self, simulation_id: str) -> dict[str, Any]:
        """GET /v1/simulations/{simulation_id}/history"""
        response = await self._client.get(f"/v1/simulations/{simulation_id}/history")
        return response.json()

File: swift/Sources/Ensoul/Resources/Simulations.swift
/// Simulations resource for the Ensoul Swift SDK.
///
/// Wraps all `/v1/simulations` endpoints, including lifecycle control
/// (start / pause / stop), SSE streaming, and event history.
///
/// Example:
/// ```swift
/// let sim = try await client.simulations.create(name: "Run 1", domainId: "acme")
/// try await client.simulations.start(sim.id, ticks: 100)
///
/// let stream = client.simulations.stream(sim.id)
/// for try await event in stream {
///     print(event.data)
/// }
/// ```
import Foundation

// MARK: - Simulations

@available(iOS 15.0, macOS 12.0, *)
public class Simulations {
    private let client: HTTPClient

    public init(client: HTTPClient) {
        self.client = client
    }

    // MARK: - Create

    /// POST /v1/simulations
    ///
    /// Creates a new simulation in the given domain. The simulation begins in a
    /// `pending` state; call `start(_:ticks:)` to begin execution.
    public func create(
        name: String,
        domainId: String,
        description: String? = nil,
        config: [String: Any]? = nil,
        participantPersonaIds: [String]? = nil
    ) async throws -> SimulationDetailResponse {
        var body: [String: Any] = [
            "name": name,
            "domain_id": domainId,
        ]
        if let description { body["description"] = description }
        if let config { body["config"] = config }
        if let participantPersonaIds { body["participant_persona_ids"] = participantPersonaIds }

        let (data, _) = try await client.post("/v1/simulations", body: body)
        let decoder = JSONDecoder()
        return try decoder.decode(SimulationDetailResponse.self, from: data)
    }

    // MARK: - Get

    /// GET /v1/simulations/{simulationId}
    public func get(_ simulationId: String) async throws -> SimulationDetailResponse {
        let (data, _) = try await client.get("/v1/simulations/\(simulationId)")
        let decoder = JSONDecoder()
        return try decoder.decode(SimulationDetailResponse.self, from: data)
    }

    // MARK: - List

    /// GET /v1/simulations
    ///
    /// Returns a paginated list of simulations as raw JSON dictionaries
    /// (the simulation summary schema varies by domain configuration).
    public func list(
        page: Int = 1,
        perPage: Int = 20
    ) async throws -> RawPage {
        let params: [String: String] = [
            "page": String(page),
            "per_page": String(perPage),
        ]
        let (data, _) = try await client.get("/v1/simulations", params: params)
        return try RawPage.from(data: data)
    }

    // MARK: - Start

    /// POST /v1/simulations/{simulationId}/start
    ///
    /// Starts or resumes a simulation. Pass `ticks` to run for a fixed number
    /// of simulation ticks before automatically pausing.
    public func start(_ simulationId: String, ticks: Int? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [:]
        if let ticks { body["ticks"] = ticks }

        let (data, _) = try await client.post(
            "/v1/simulations/\(simulationId)/start",
            body: body
        )
        return try Self.jsonObject(from: data)
    }

    // MARK: - Pause

    /// POST /v1/simulations/{simulationId}/pause
    ///
    /// Pauses a running simulation. The simulation can be resumed with `start`.
    public func pause(_ simulationId: String) async throws -> [String: Any] {
        let (data, _) = try await client.post(
            "/v1/simulations/\(simulationId)/pause",
            body: [String: Any]()
        )
        return try Self.jsonObject(from: data)
    }

    // MARK: - Stop

    /// POST /v1/simulations/{simulationId}/stop
    ///
    /// Permanently stops a simulation. This cannot be undone; use `pause` if you
    /// intend to resume later.
    public func stop(_ simulationId: String) async throws -> [String: Any] {
        let (data, _) = try await client.post(
            "/v1/simulations/\(simulationId)/stop",
            body: [String: Any]()
        )
        return try Self.jsonObject(from: data)
    }

    // MARK: - Stream

    /// GET /v1/simulations/{simulationId}/stream
    ///
    /// Opens an SSE stream that emits real-time simulation events. Returns an
    /// `SSEStream` that the caller iterates with `for try await event in stream { }`.
    ///
    /// - Note: This method is synchronous (non-throwing) — no network request is
    ///   sent until the caller begins consuming the stream.
    public func stream(_ simulationId: String) -> SSEStream {
        return client.streamSSE(
            method: "GET",
            path: "/v1/simulations/\(simulationId)/stream",
            body: nil
        )
    }

    // MARK: - Get Events

    /// GET /v1/simulations/{simulationId}/events
    ///
    /// Returns a paginated list of historical simulation events.
    /// Events are returned as raw dictionaries because their shape is
    /// domain-specific (e.g. interaction events, state snapshots, etc.).
    public func getEvents(
        _ simulationId: String,
        page: Int = 1,
        perPage: Int = 20
    ) async throws -> RawPage {
        let params: [String: String] = [
            "page": String(page),
            "per_page": String(perPage),
        ]
        let (data, _) = try await client.get(
            "/v1/simulations/\(simulationId)/events",
            params: params
        )
        return try RawPage.from(data: data)
    }

    // MARK: - Get History

    /// GET /v1/simulations/{simulationId}/history
    ///
    /// Returns the full aggregated history of a simulation as a raw JSON dict.
    /// The history schema is domain-specific.
    public func getHistory(_ simulationId: String) async throws -> [String: Any] {
        let (data, _) = try await client.get("/v1/simulations/\(simulationId)/history")
        return try Self.jsonObject(from: data)
    }

    // MARK: - Private helpers

    /// Decode response `Data` as a top-level JSON object.
    private static func jsonObject(from data: Data) throws -> [String: Any] {
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw EnsoulAPIError(
                statusCode: 200,
                error: "ParseError",
                message: "Expected JSON object in simulation response"
            )
        }
        return dict
    }
}
